### PR TITLE
Recorder participant [16665]

### PR DESCRIPTION
**Pull request by jparisu**
After #1

### CODE REVIEW

File: ddsrecorder/src/cpp/core/ParticipantFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        case ParticipantKind::blank:
+            // BlankParticipant
+            return std::make_shared<BlankParticipant>(participant_configuration->id);
+
+        case ParticipantKind::dummy:
+            // DummyParticipant
+            return std::make_shared<DummyParticipant>(participant_configuration, payload_pool, discovery_database);
+
+        case ParticipantKind::recorder:
+            // DummyParticipant
```
Comment (juanlofer-eprosima) [inline ddsrecorder/src/cpp/core/ParticipantFactory.cpp:61]:
```suggestion
            // Recorder Participant
```

File: ddsrecorder/src/cpp/participant/implementations/recorder/RecorderParticipant.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * TODO(recorder)
+ */
+class RecorderParticipant : public BaseParticipant
+{
+public:
+
+    /**
+     * @brief Construct a new Dummy Participant object
```
Comment (juanlofer-eprosima) [inline ddsrecorder/src/cpp/participant/implementations/recorder/RecorderParticipant.hpp:39]:
```suggestion
     * @brief Construct a new Recorder Participant object
```

File: ddsrecorder/src/cpp/participant/implementations/recorder/RecorderParticipant.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+     * Apart from BaseParticipant, it adds this new object to a static variable so it could be reached from outside
+     * the DDSRouter.
+     */
+    RecorderParticipant(
+            std::shared_ptr<configuration::ParticipantConfiguration> participant_configuration,
+            std::shared_ptr<PayloadPool> payload_pool,
+            std::shared_ptr<DiscoveryDatabase> discovery_database);
+
+    /**
+     * @brief Destroy the Dummy Participant object
```
Comment (juanlofer-eprosima) [inline ddsrecorder/src/cpp/participant/implementations/recorder/RecorderParticipant.hpp:51]:
```suggestion
     * @brief Destroy the Recorder Participant object
```

File: ddsrecorder/src/cpp/reader/implementations/auxiliar/InternalReader.hpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+     *
+     * @param data : oldest data to take
+     * @return \c RETCODE_OK if data has been correctly taken
+     * @return \c RETCODE_NO_DATA if \c data_to_send_ is empty
+     */
+    utils::ReturnCode take_(
+            std::unique_ptr<types::DataReceived>& data) noexcept override;
+
+    //! Stores the data that must be retrieved with \c take() method
+    using DataReceviedType = utils::Atomicable<std::queue<std::unique_ptr<types::DataReceived>>>;
```
Comment (juanlofer-eprosima) [inline ddsrecorder/src/cpp/reader/implementations/auxiliar/InternalReader.hpp:67]:
```suggestion
    using DataReceivedType = utils::Atomicable<std::queue<std::unique_ptr<types::DataReceived>>>;
```

File: ddsrecorder/src/cpp/participant/implementations/recorder/RecorderParticipant.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    // Simulate that there is a reader of type object to force this track creation
+    discovery_database_->add_endpoint(
+        simulate_endpoint_(recorder::type_object_topic())
+    );
+
+    // Force for every topic found to create track by creating simulated readers
+    // NOTE: this could change for: in DDSRouter change that only readers create track
+    discovery_database_->add_endpoint_discovered_callback(
+        [this](types::Endpoint endpoint_discovered){
+            if (endpoint_discovered.discoverer_participant_id() != this->id())
```
Comment (juanlofer-eprosima) [inline ddsrecorder/src/cpp/participant/implementations/recorder/RecorderParticipant.cpp:48]:
```suggestion
            if (endpoint_discovered.is_writer() && endpoint_discovered.discoverer_participant_id() != this->id())
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Review by rsanchez15 — CHANGES_REQUESTED** [review]
Change ddsrecorder.repos to pull master/main branches

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Review by rsanchez15 — APPROVED** [review]

**Comment by Danipiza** [comment]
I added "#2" in the flaky fix PR trying to mention the second flaky tests, but it mentioned this PR

**Pull request merged**